eth.ts

```diff
@@ -2,30 +2,69 @@ import Web3 from "web3";
 import {compiledContracts} from "./compiled-contracts";
 import { Contract as Web3Contract } from "web3-eth-contract";
 import BN from "bn.js";
+import { Contracts } from "./typings/contracts";
 const HDWalletProvider = require("truffle-hdwallet-provider");
 
 export const ETHEREUM_URL = process.env.ETHEREUM_URL || "http://localhost:7545";
 const ETHEREUM_MNEMONIC = process.env.ETHEREUM_MNEMONIC || "vanish junk genuine web seminar cook absurd royal ability series taste method identify elevator liquid";
 
+export const defaultWeb3Provider = () => new Web3(new HDWalletProvider(
+    ETHEREUM_MNEMONIC,
+    ETHEREUM_URL,
+    0,
+    100,
+    false
+    ));
 
-const refreshWeb3 = () => {
-    web3 = new Web3(new HDWalletProvider(
-        ETHEREUM_MNEMONIC,
-        ETHEREUM_URL,
-        0,
-        100,
-        false
-        )
-    );
-};
+export class Web3Driver{
+    private web3 : Web3;
+    private nonceVersion = 0;
+    constructor(private web3Provider : () => Web3 = defaultWeb3Provider){
+        this.web3 = this.web3Provider();
+    }
 
-export var web3;
-refreshWeb3();
+    get eth(){
+        return this.web3.eth;
+    }
+    get currentProvider(){
+        return this.web3.currentProvider;
+    }
+
+    async deploy<N extends keyof Contracts>(contractName: N, args: any[], options?: any) {
+        try {
+            const abi = compiledContracts[contractName].abi;
+            const accounts = await this.web3.eth.getAccounts();
+            let web3Contract = await new this.web3.eth.Contract(abi).deploy({
+                data: compiledContracts[contractName].bytecode,
+                arguments: args || []
+            }).send({
+                from: accounts[0],
+                ...(options || {})
+            });
+            let lastNonce = this.nonceVersion;
+            const web3ContractProvider = () => {
+                if (this.nonceVersion != lastNonce){
+                    web3Contract = new this.web3.eth.Contract(abi, web3Contract.options.address);
+                }
+                return web3Contract;
+            }
+            return new Contract(this, abi, web3ContractProvider) as Contracts[N];
+        } catch (e) {
+            this.refreshNonce();
+            throw e;
+        }
+    }
+    
+    refreshNonce(){
```

Comment:
i agree the names are confusing here. this function refreshes the `web3` object, because of a stale nonce. it's not directly refreshing the nonce. and it is doing something far more meaningful that just refreshing the nonce.

Reply:
just `refresh` than ?